src/coroutine.rs

```diff
@@ -26,13 +26,44 @@ pub use cancel::CancelHandle;
 
 const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
 
+trait CoroutineFuture {
+    fn poll(
+        self: Pin<&mut Self>,
+        py: Python<'_>,
+        waker: &Waker,
+        allow_threads: bool,
+    ) -> Poll<PyResult<PyObject>>;
+}
+
+impl<F, T, E> CoroutineFuture for F
+where
+    F: Future<Output = Result<T, E>> + Send,
+    T: IntoPy<PyObject> + Send,
+    E: Into<PyErr> + Send,
+{
+    fn poll(
+        self: Pin<&mut Self>,
+        py: Python<'_>,
+        waker: &Waker,
+        allow_threads: bool,
+    ) -> Poll<PyResult<PyObject>> {
+        let result = if allow_threads {
+            py.allow_threads(|| self.poll(&mut Context::from_waker(waker)))
```

Comment:
Is it important the the context is constructed without the GIL being held? Could this factored out into a binding to make it more obvious that this is the same between both cases, e.g. `let cx = &mut Context::from_waker(waker);`?

Reply:
`Context` is explicitely not `Send`